// openapi-framework/dist/src/types.js
"use strict";
exports.__esModule = true;
var ConsoleDebugAdapterLogger = /** @class */ (function () {
    function ConsoleDebugAdapterLogger() {
    }
    /**
     * `console.debug` is just an alias for `.log()`, and we want debug logging to be optional.
     * This class delegates to `console` and overrides `.debug()` to be a no-op.
     */
    ConsoleDebugAdapterLogger.prototype.debug = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        // no-op
    };
    ConsoleDebugAdapterLogger.prototype.error = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        console.error.apply(console, [message].concat(optionalParams));
    };
    ConsoleDebugAdapterLogger.prototype.info = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        console.info.apply(console, [message].concat(optionalParams));
    };
    ConsoleDebugAdapterLogger.prototype.trace = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        console.trace.apply(console, [message].concat(optionalParams));
    };
    ConsoleDebugAdapterLogger.prototype.warn = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        console.warn.apply(console, [message].concat(optionalParams));
    };
    return ConsoleDebugAdapterLogger;
}());
exports.ConsoleDebugAdapterLogger = ConsoleDebugAdapterLogger;
