// rabbitmq-server/index.js
module.exports = start;

const fs = require('fs')
const ini = require('ini')
const { fork, spawn } = require('child_process');
var bindir = "";
var rootdir = "";

const start_options = {
  slient:true,
  windowsHide: true,
  detached:true,
    stdio: [null, null, null, 'ipc']
};

function update_ini(filepath){
	console.log("Checking config file: "+filepath)
	var config = ini.parse(fs.readFileSync(filepath,"utf-8"))
	console.log(config)
	if (bindir != config["erlang"]["Bindir"]){
		config["erlang"]["Bindir"] = bindir;
		console.log("Detected installation moved, updating paths in file " + filepath)
		fs.writeFileSync(filepath, ini.stringify(config))
	}
	if (rootdir != config["erlang"]["Rootdir"]){
		config["erlang"]["Rootdir"] = rootdir;
		console.log("Detected installation moved, updating paths in file " + filepath)
		fs.writeFileSync(filepath, ini.stringify(config))
	}
	
}
function start(){
	//check and set needed erlang ini settings to absolute paths
	bindir = __dirname + "/bin/erl10.5/erts-10.5/bin/";
	bindir = bindir.split("\\").join("/")
	rootdir = __dirname + "/bin/erl10.5/";
	rootdir = rootdir.split("\\").join("/")
	var erl_ini = rootdir + "/bin/erl.ini";
	erl_ini.split("\\").join("/")
	update_ini(erl_ini)
	var erts_ini = rootdir + "erts-10.5/bin/erl.ini";
	erts_ini = erts_ini.split("\\").join("/")
	update_ini(erts_ini)
	//set environment variable
	start_options["env"]={"ERLANG_HOME":rootdir}
	child = spawn(__dirname + "/bin/erl10.5/lib/rabbitmq_server-3.8.3/sbin/rabbitmq-server.bat", ["-detached"], start_options);

	child.on('message', (data) => {
		console.log("Exiting RabbitMQ msg:" + data);
		child.unref();
		process.exit(0);
	});

	process.on('SIGTERM', () => {
	  console.log('SIGTERM signal received.');
	  //child.kill('SIGINT');
	});

	process.on('SIGINT', () => {
	  console.log('SIGINT signal received.');
	  //child.kill('SIGINT');
	});
}
