// app-builder/lib/cjs/app-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.functionList = exports.compose = exports.AppBuilder = void 0;
const compose_js_1 = require("./compose.js");
Object.defineProperty(exports, "compose", { enumerable: true, get: function () { return compose_js_1.compose; } });
Object.defineProperty(exports, "functionList", { enumerable: true, get: function () { return compose_js_1.functionList; } });
class AppBuilder {
    constructor() {
        this.middleware = [];
    }
    build() {
        if (!this.middleware.length) {
            throw new Error('Usage error: must have at least one middleware');
        }
        return compose_js_1.compose(this.middleware);
    }
    use(mw) {
        if ('function' !== typeof mw) {
            throw new TypeError(`${mw}, must be a middleware function accpeting (context, next) arguments`);
        }
        this.middleware.push(mw);
        return this;
    }
}
exports.AppBuilder = AppBuilder;
function createAppBuilder() {
    return new AppBuilder();
}
exports.default = createAppBuilder;
