// resolve-dependencies/lib/loader.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Loader = void 0;
const path_1 = require("path");
const node_builtins_1 = __importDefault(require("./node-builtins"));
const loader = __importStar(require("./node-loader"));
const file_1 = require("./file");
class Loader {
    constructor(options) {
        this.options = options;
    }
    loadEntry(workingDirectory, request, files = {}, warnings = []) {
        const mainFile = (0, file_1.ensureDottedRelative)(workingDirectory, (0, path_1.resolve)(workingDirectory, request));
        const entry = this.load(workingDirectory, mainFile, files, warnings);
        return { entry: entry, files, warnings };
    }
    load(cd, request, files = {}, warnings, context) {
        var _a;
        const options = { ...this.options };
        if (context) {
            Object.assign(options, { context });
        }
        const file = loader.load(cd, request, options);
        if ('warning' in file) {
            warnings.push(file.warning);
            return null;
        }
        if (files[file.absPath] !== undefined) {
            return files[file.absPath];
        }
        else {
            files[file.absPath] = file;
        }
        const packageGlobs = [];
        if (file.moduleRoot && (0, file_1.hasModuleGlobs)(file)) {
            packageGlobs.push(...(0, file_1.nodeModuleGlobs)(file, false));
        }
        if (!packageGlobs.length && (0, file_1.hasModuleGlobs)((file === null || file === void 0 ? void 0 : file.belongsTo) || {})) {
            packageGlobs.push(...(0, file_1.nodeModuleGlobs)(file.belongsTo || {}, false));
        }
        const fileDir = (0, path_1.dirname)(file.absPath), ctx = {
            moduleRoot: file.moduleRoot || ((_a = file.belongsTo) === null || _a === void 0 ? void 0 : _a.moduleRoot) || undefined,
            package: file.package,
            expanded: Boolean(file.contextExpanded),
            globs: packageGlobs,
        }, rejectBuiltins = (req) => {
            if (~node_builtins_1.default.indexOf(req)) {
                file.deps[req] = null;
                return false;
            }
            return true;
        };
        Object.keys(file.deps)
            .sort()
            .filter(rejectBuiltins)
            .map((req) => {
            const dep = (file.deps[req] = this.load(fileDir, req, files, warnings, ctx));
            if (dep) {
                dep.belongsTo = file.moduleRoot ? file : file.belongsTo;
            }
        });
        return file;
    }
}
exports.Loader = Loader;
